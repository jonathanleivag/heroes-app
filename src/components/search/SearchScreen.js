import React, { useMemo } from 'react'
import queryString from 'query-string'
import { Row, Col, Button, Form, Alert } from 'react-bootstrap'
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../heroes/HeroCard'
import { useLocation } from 'react-router-dom'
import { getHeroByName } from '../selectors/getHeroByName'

export const SearchScreen = ({ history: { push } }) => {

    const location = useLocation()
    const { q = '' } = queryString.parse(location.search)
    const [values, handlerInputChange] = useForm({ searchText: q })
    const { searchText } = values
    const heroesFiltered = useMemo(() => getHeroByName(q), [q])
    const handlerSearch = event => {
        event.preventDefault()
        push(`?q=${searchText}`)
    }
    return (
        <div>
            <Row>
                <Col md={5}>
                    <h1>Search Screen</h1>
                    <hr />
                    <Form autoComplete='off' onSubmit={handlerSearch}>
                        <Form.Group controlId='inpControl'>
                            <Form.Label>Search:</Form.Label>
                            <Form.Control type='text' name='searchText' value={searchText} onChange={handlerInputChange} placeholder='Search heroes' />
                        </Form.Group>
                        <Button variant="outline-info" size='lg' block type="submit">
                            Search
                        </Button>
                    </Form>
                </Col>
                <Col md={7}>
                    <h2 style={{ fontSize: '2.5rem' }}>Result</h2>
                    <hr />
                    {q === '' && <Alert className='mt-5' variant='info'>Search a hero</Alert>}
                    {(q !== '' && heroesFiltered.length === 0) && <Alert className='mt-5' variant='danger'>There is no a hero with {q} </Alert>}
                    {heroesFiltered.map(hero => <HeroCard key={hero.id} {...hero} />)}
                </Col>
            </Row>

        </div>
    )
}