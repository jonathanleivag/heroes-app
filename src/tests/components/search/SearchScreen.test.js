import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import '@testing-library/jest-dom'
import { SearchScreen } from '../../../components/search/SearchScreen';
import { act } from '@testing-library/react';

describe('Pruebas en <SearchScreen />', () => {
    test('debe de mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route exact path='/search' component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot()
    })

    test('debe de mostrar a Batman y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Spider%20Man']}>
                <Route exact path='/search' component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper.find('input').prop('value')).toBe('Spider Man')
        expect(wrapper).toMatchSnapshot()
    })

    test('debe de mostrar un error si no se encuentra el Hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Spider%20Mannn']}>
                <Route exact path='/search' component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper.find('input').prop('value')).toBe('Spider Mannn')
        expect(wrapper.find('Alert').text().trim()).toBe('There is no a hero with Spider Mannn')
        expect(wrapper).toMatchSnapshot()
    })

    test('debe de llamar el push del history', () => {
        const history = {
            push: jest.fn()
        };
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Spider%20Man']}>
                <Route exact path='/search' component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        )
        act(()=> wrapper.find('input').prop('onChange')({ target: { name: 'searchText', value: 'Spider Man' } }))    
        wrapper.find('form').prop('onSubmit')({ preventDefault() { } })
        expect(history.push).toHaveBeenCalledWith(`?q=Spider Man`)
    })

})
