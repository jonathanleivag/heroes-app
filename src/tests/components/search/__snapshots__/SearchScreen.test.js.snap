// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-md-5"
    >
      <h1>
        Search Screen
      </h1>
      <hr />
      <form
        autoComplete="off"
        className=""
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="inpControl"
          >
            Search:
          </label>
          <input
            className="form-control"
            id="inpControl"
            name="searchText"
            onChange={[Function]}
            placeholder="Search heroes"
            type="text"
            value="Spider Man"
          />
        </div>
        <button
          className="btn btn-outline-info btn-block btn-lg"
          disabled={false}
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-md-7"
    >
      <h2
        style={
          Object {
            "fontSize": "2.5rem",
          }
        }
      >
        Result
      </h2>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Spider Man"
              className="card-img"
              src="./assets/images/marvel-spider.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Spider Man
              </h5>
              <p
                className="card-text"
              >
                Peter Parker
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Amazing Fantasy #15
                   
                </small>
              </p>
              <a
                href="/hero/marvel-spider"
                onClick={[Function]}
              >
                ...MÃ¡s
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrar un error si no se encuentra el Hero 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-md-5"
    >
      <h1>
        Search Screen
      </h1>
      <hr />
      <form
        autoComplete="off"
        className=""
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="inpControl"
          >
            Search:
          </label>
          <input
            className="form-control"
            id="inpControl"
            name="searchText"
            onChange={[Function]}
            placeholder="Search heroes"
            type="text"
            value="Spider Mannn"
          />
        </div>
        <button
          className="btn btn-outline-info btn-block btn-lg"
          disabled={false}
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-md-7"
    >
      <h2
        style={
          Object {
            "fontSize": "2.5rem",
          }
        }
      >
        Result
      </h2>
      <hr />
      <div
        className="fade mt-5 alert alert-danger show"
        role="alert"
      >
        There is no a hero with 
        Spider Mannn
         
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-md-5"
    >
      <h1>
        Search Screen
      </h1>
      <hr />
      <form
        autoComplete="off"
        className=""
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="inpControl"
          >
            Search:
          </label>
          <input
            className="form-control"
            id="inpControl"
            name="searchText"
            onChange={[Function]}
            placeholder="Search heroes"
            type="text"
            value=""
          />
        </div>
        <button
          className="btn btn-outline-info btn-block btn-lg"
          disabled={false}
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-md-7"
    >
      <h2
        style={
          Object {
            "fontSize": "2.5rem",
          }
        }
      >
        Result
      </h2>
      <hr />
      <div
        className="fade mt-5 alert alert-info show"
        role="alert"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
